File: streampark-flink/streampark-flink-packer/src/main/scala/org/apache/streampark/flink/packer/pipeline/impl/FlinkYarnApplicationBuildPipeline.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.packer.pipeline.impl

import org.apache.streampark.common.conf.Workspace
import org.apache.streampark.common.enums.DevelopmentMode
import org.apache.streampark.common.fs.{FsOperator, HdfsOperator, LfsOperator}
import org.apache.streampark.common.util.Utils
import org.apache.streampark.flink.packer.maven.MavenTool
import org.apache.streampark.flink.packer.pipeline._

import org.apache.commons.codec.digest.DigestUtils

import java.io.{File, FileInputStream, IOException}

/** Building pipeline for flink yarn application mode */
class FlinkYarnApplicationBuildPipeline(request: FlinkYarnApplicationBuildRequest)
  extends BuildPipeline {

  /** the type of pipeline */
  override def pipeType: PipelineType = PipelineType.FLINK_YARN_APPLICATION

  override def offerBuildParam: FlinkYarnApplicationBuildRequest = request

  /**
   * the actual build process. the effective steps progress should be implemented in multiple
   * BuildPipeline.execStep() functions.
   */
  @throws[Throwable]
  override protected def buildProcess(): SimpleBuildResponse = {
    execStep(1) {
      request.developmentMode match {
        case DevelopmentMode.FLINK_SQL =>
          LfsOperator.mkCleanDirs(request.localWorkspace)
          HdfsOperator.mkCleanDirs(request.yarnProvidedPath)
        case _ =>
      }
      logInfo(s"recreate building workspace: ${request.yarnProvidedPath}")
    }.getOrElse(throw getError.exception)

    val mavenJars =
      execStep(2) {
        request.developmentMode match {
          case DevelopmentMode.FLINK_SQL =>
            val mavenArts = MavenTool.resolveArtifacts(request.dependencyInfo.mavenArts)
            mavenArts.map(_.getAbsolutePath) ++ request.dependencyInfo.extJarLibs
          case _ => Set[String]()
        }
      }.getOrElse(throw getError.exception)

    execStep(3) {
      mavenJars.foreach(
        jar => {
          uploadToHdfs(FsOperator.lfs, jar, request.localWorkspace)
          uploadToHdfs(FsOperator.hdfs, jar, request.yarnProvidedPath)
        })
    }.getOrElse(throw getError.exception)

    SimpleBuildResponse()
  }

  @throws[IOException]
  private[this] def uploadToHdfs(fsOperator: FsOperator, origin: String, target: String): Unit = {
    val originFile = new File(origin)
    if (!fsOperator.exists(target)) {
      fsOperator.mkdirs(target)
    }
    if (originFile.isFile) {
      // check file in upload dir
      fsOperator match {
        case FsOperator.lfs =>
          fsOperator.copy(originFile.getAbsolutePath, target)
        case FsOperator.hdfs =>
          val uploadFile = s"${Workspace.remote.APP_UPLOADS}/${originFile.getName}"
          if (fsOperator.exists(uploadFile)) {
            Utils.using(new FileInputStream(originFile))(
              inputStream => {
                if (DigestUtils.md5Hex(inputStream) != fsOperator.fileMd5(uploadFile)) {
                  fsOperator.upload(originFile.getAbsolutePath, uploadFile)
                }
              })
          } else {
            fsOperator.upload(originFile.getAbsolutePath, uploadFile)
          }
          // copy jar from upload dir to target dir
          fsOperator.copy(uploadFile, target)
      }
    } else {
      fsOperator match {
        case FsOperator.hdfs => fsOperator.upload(originFile.getAbsolutePath, target)
        case _ =>
      }
    }
  }

}

object FlinkYarnApplicationBuildPipeline {
  def of(request: FlinkYarnApplicationBuildRequest): FlinkYarnApplicationBuildPipeline =
    new FlinkYarnApplicationBuildPipeline(request)
}

File: streampark-flink/streampark-flink-packer/src/main/java/org/apache/streampark/flink/packer/pipeline/impl/FlinkYarnApplicationBuildPipeline.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.packer.pipeline.impl;

import org.apache.streampark.common.enums.FlinkJobType;
import org.apache.streampark.common.fs.FsOperator;
import org.apache.streampark.flink.packer.maven.Artifact;
import org.apache.streampark.flink.packer.maven.DependencyInfo;
import org.apache.streampark.flink.packer.maven.MavenTool;
import org.apache.streampark.flink.packer.pipeline.BuildPipeline;
import org.apache.streampark.flink.packer.pipeline.FlinkYarnApplicationBuildRequest;
import org.apache.streampark.flink.packer.pipeline.PipelineTypeEnum;
import org.apache.streampark.flink.packer.pipeline.SimpleBuildResponse;
import org.apache.streampark.flink.packer.pipeline.YarnJarUploader;

import java.io.File;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

/** Building pipeline for flink yarn application mode */
public class FlinkYarnApplicationBuildPipeline extends BuildPipeline {

    private final FlinkYarnApplicationBuildRequest request;

    public FlinkYarnApplicationBuildPipeline(FlinkYarnApplicationBuildRequest request) {
        this.request = request;
    }

    @Override
    public PipelineTypeEnum pipeType() {
        return PipelineTypeEnum.FLINK_YARN_APPLICATION;
    }

    @Override
    public FlinkYarnApplicationBuildRequest offerBuildParam() {
        return request;
    }

    @Override
    public SimpleBuildResponse buildProcess() {
        boolean sqlMode =
            request.flinkJobType() == FlinkJobType.FLINK_SQL
                || request.flinkJobType() == FlinkJobType.PYFLINK;
        DependencyInfo dependencyInfo = request.dependencyInfo();
        if (request.flinkJobType() == FlinkJobType.FLINK_SQL) {
            Set<String> extJarLibs = new HashSet<>(dependencyInfo.extJarLibs());
            String appHome = System.getProperty("app.home", "/streampark");
            File snakeyaml = new File(appHome, "lib/snakeyaml-2.0.jar");
            if (snakeyaml.isFile()) {
                extJarLibs.add(snakeyaml.getAbsolutePath());
            } else {
                try {
                    MavenTool.resolveArtifacts(
                        Collections.singleton(new Artifact("org.yaml", "snakeyaml", "2.0")))
                        .stream()
                        .map(File::getAbsolutePath)
                        .forEach(extJarLibs::add);
                } catch (Exception e) {
                    throw new IllegalStateException("Failed to resolve snakeyaml for Flink SQL yarn application", e);
                }
            }
            dependencyInfo = new DependencyInfo(dependencyInfo.mavenArts(), extJarLibs);
            buildAndUploadSqlFatJar();
        }
        runYarnSqlBuildSteps(
            request.localWorkspace(),
            request.yarnProvidedPath(),
            sqlMode,
            dependencyInfo);
        return new SimpleBuildResponse();
    }

    private void buildAndUploadSqlFatJar() {
        try {
            String shadedJarOutputPath = request.getShadedJarPath(request.localAppHome());
            File jar =
                MavenTool.buildFatJar(
                    request.mainClass(),
                    request.providedLibs(),
                    shadedJarOutputPath);
            logInfo("Output shaded flink SQL jar: " + jar.getAbsolutePath());
            YarnJarUploader.uploadJarToHdfsOrLfs(
                FsOperator.hdfs(), jar.getAbsolutePath(), request.remoteAppHome());
        } catch (Exception e) {
            throw new IllegalStateException("Failed to build Flink SQL fat jar for yarn application", e);
        }
    }

    public static FlinkYarnApplicationBuildPipeline of(FlinkYarnApplicationBuildRequest request) {
        return new FlinkYarnApplicationBuildPipeline(request);
    }
}
